Source/DafnyLanguageServer/Language/DiagnosticErrorReporter.cs

```diff
@@ -35,9 +35,13 @@ public class DiagnosticErrorReporter : ErrorReporter {
     public IReadOnlyList<Diagnostic> GetDiagnostics(DocumentUri documentUri) {
       rwLock.EnterReadLock();
       try {
+        var alternativeUntitled = documentUri.GetFileSystemPath();
```

Comment:
What is happening here is that, when we receive verification errors, we call this method to get a DocumentURI from the token (PathExtensions.cs)

```
   public static DocumentUri GetDocumentUri(this Boogie.IToken token) {
      return DocumentUri.FromFileSystemPath(token.filename);
    }
```

Originally, when token were parsed, they are given a string for a fileName that is obtained through the following way (PathExtensions.cs)
```
private static string GetFilePath(DocumentUri uri) {
      // GetFileSystemPath() is used since Path resolves to a non-Windows path format on Windows, e.g.:
      // /d:/data/file.dfy
      return uri.GetFileSystemPath();
    }
```

It just happens that this roundtrip from the DocumentUri provided by the language server does not create the same DocumentUri if the original DocumentUri is `DocumentUri.parse("untitled:Untitled-1")` because the scheme is different.

As a follow-up to your idea, I tried to replace the roundtrip with `.ToString()` and `Parse()`, but suddenly without my change tests in DefinitionTest and HoverTest such as d`DefinitionInConstructorInvocationOfUserDefinedTypeOfForeignFileReturnsLinkToForeignFile` are failing and I'm not sure why.

Since the change I make is conservative (it will only trigger if the key is not in the dictionary), I'm pretty confident it will solve the problem for now.
So for now, I'd vote to use it.

Reply:
Shouldn't we then change `GetDocumentUri(this Boogie.IToken token)` so it returns a DocumentURI with an "untitled" scheme for unsaved files?

Reply:
The problem is, how do you detect that a file is untitled? I wouldn't bet my life on `tok.fileName.StartsWith("untitled:"`, because I don't know the URI scheme enough (can it be different in different languages? Is it just because there is a word and a colon?)
I think the current solution is good enough.

Reply:
So.. you're saying `GetDiagnostics` receives a `DocumentUri` which is broken in the sense that it refers to an untitled document, but that is not in the scheme, and we can fix that DocumentUri by printing and parsing again.

Should we make it so that `GetDiagnostics` doesn't receive broken parameters?

Reply:
If we fixed the argument, we would need to replicate this code for every call to `GetDiagnostics`, and I prefer to avoid duplication.

Because this is a regression issue, I dug with `git bisect` to see when untitled documents failed to display diagnostics. The commit introducing this (previously non-tested) regression issue is the one created by PR #2093. More specifically, it probably has to do with the removal of [this code](https://github.com/dafny-lang/dafny/pull/2093/files#diff-937b4809390419fc4ebbcb08e0dee41c7318f8e6e4e34fe22cad29981387b38aL48) that was relying on `DiagnosticErrorReporter` which was using `GetDocumentUriOrDefault(error.Tok)`. I don't know what went wrong there.

Since you were the author of that (very recent) PR, if you have other experimental ideas in mind like but you are not sure about them, feel free to checkout this PR, remove the alternative URI computation, and implement another fix. You might have more insights than I. I'd be happy to approve your follow-up PR.

If you decide it's not worth your immediate time, let's just merge this PR. Opening untitled documents is my typical workflow when developing UI features, and this issue has increasingly annoyed me a lot for the last 8 days.
You could always push a better fix for your PR later.

Reply:
> Since you were the author of that (very recent) PR, if you have other experimental ideas in mind like but you are not sure about them, feel free to checkout this PR, remove the alternative URI computation, and implement another fix. You might have more insights than I. I'd be happy to approve your follow-up PR.

If you decide it's not worth your immediate time, let's just merge this PR. Opening untitled documents is my typical workflow when developing UI features, and this issue has increasingly annoyed me a lot for the last 8 days.
You could always push a better fix for your PR later.

That sounds reasonable :) sorry for your discomfort